### PR TITLE
CRM-20905: Make (max emails kill switch) a configuration instead of hardcoded value

### PR DESCRIPTION
**Pull request by omarabuhussein**
## Problem
Currently , when trying to send a direct to more than one contacts (direct means without using civimail ) there is a limit to maximum 50 contacts . 

2017-07-18 15_37_54-find contacts _ civicrm sandbox on drupal

and there is no way to change this even if your server configurations allow more or less than this limit. 

## Solution

I changed this to be a configuration (with machine name : **max_emails_kill_switch**) that you can find inside (Administer >> System Settings >> Misc Undelete, PDFs, Limits, Logging, Captcha, etc.) page  .

screenshot-dmaster2 local-2017-07-18-13-46-54

2017-07-18 15_48_30-find contacts _ dmaster2

---

 * [CRM-20905: Make max emails kill switch a config instead of hardcoded value set to 50](https://issues.civicrm.org/jira/browse/CRM-20905)

### CODE REVIEW

**Comment by mattwire** [comment]
Can you add a description for the setting, like the other ones on that screen?

**Comment by omarabuhussein** [comment]
sure @mattwire

**Comment by seamuslee001** [comment]
Jenkins ok to test

File: CRM/Contact/Task.php [hunk 1]
```diff
@@ -182,6 +176,14 @@ public static function initTasks() {
         ),
       );
 
+      $maxEmailKillSwitch = Civi::settings()->get('max_emails_kill_switch');
+      self::$_tasks[self::EMAIL_CONTACTS] = array(
+        'title' => ts('Email - send now (to ' . $maxEmailKillSwitch . ' or less)'),
```
Comment (colemanw) [inline CRM/Contact/Task.php]: You can't concatenate variables directly into the `ts()` function. Instead, do 

    ts('Email - send now (to %1 or less)', array(1 => $maxEmailKillSwitch))

**Review by colemanw — CHANGES_REQUESTED** [review]

**Comment by eileenmcnaughton** [comment]
Frig - I've been working on the same thing without having spotted this!

**Comment by eileenmcnaughton** [comment]
I'm going to still push up my version to PRs & then we can converge on an agreed version. The first difference I note is that I treated it as a 'Mailing' setting rather than a miscellaneous (& have put up a refactor of that form to make it easy to add to that form https://github.com/civicrm/civicrm-core/pull/10704 - that PR should be mergeable independently of where we go with this)

**Comment by eileenmcnaughton** [comment]
@omarabuhussein  this is my version https://github.com/civicrm/civicrm-core/pull/10705/files

Key differences are
1) I put it in the mail settings form
2) differences in description
3) My PR is dependent on #10704 before the setting will show up
4) I left off the validation. I think it should come from the metadata - either in the type or as a validation callback. I feel like the existing validation is inconsistent enough that could be addressed separately

**Comment by omarabuhussein** [comment]
@eileenmcnaughton will check both of your PRs and see ... but the thing is, (max emails kill switch) is not related to mailing settings ... mailing settings are for CiviMail component and they will not be available if it was disabled while (Email - Send now (to 50 or less)) search task works whither if CiviMail is disabled or not which is I why I put it in Miscellaneous section.

**Comment by omarabuhussein** [comment]
> 1. I put it in the mail settings form

I reviewed https://github.com/civicrm/civicrm-core/pull/10704 , look fine to me generaly but left view notes .. regarding https://github.com/civicrm/civicrm-core/pull/10705/files then as I pointed above this setting is not related to CiviMail component so it fit more into Misc section .

> 2.differences in description

I don't think I am that good in writing description , we could settle on yours it sounds more informative than mine

> 3. I left off the validation. I think it should come from the metadata - either in the type or as a validation callback. I feel like the existing validation is inconsistent enough that could be addressed separately

agree

**Comment by eileenmcnaughton** [comment]
@omarabuhussein it seems like we are converged - except for where to put the setting. I would rather ask that question of someone closer to the users - like @agh1 or @michaelmcandrew - my instinct is that this is about where we draw the line between CiviMail & non civimail so from a user pov they would see it in mailing settings

**Comment by colemanw** [comment]
It's a good point though that if this setting applies to the whole 
system, then it doesn't make sense to put it on a screen that disappears 
if the CiviMail component is disabled.

**Comment by eileenmcnaughton** [comment]
Ah @colemanw I hadn't thought of that! Although it's worth noting the history here - originally Lobo refused to make this setting easier to access because he was worried people would cause the CiviCRM product to have a bad spam score. I am happy with that little bit of inaccessibilty if we still think there is any risk of that

**Comment by agh1** [comment]
@colemanw and @omarabuhussein make an excellent point that it should not be in Mailing Settings.  

Personally, I'm with Lobo: even if someone is doing it for a good cause, sending a mailing to >50 people without any way to opt out goes in the category of spam tactics.  People will get punished by their hosts or recipients, and CiviCRM will be complicit.

@eileenmcnaughton tucking this setting in a place where you have to have CiviMail does have the nice benefit of requiring someone to be somewhat capable of sending bulk emails the right way, but I don't want to hedge on the decison.  Responsible people will be responsible, and spammers will enable CiviMail, set the setting, and then go about their way sending thousands of quick "send an email" messages.

On the other hand, I strongly feel that we need a moratorium on Misc.  With three other settings pages related to mailing and no mention of mail in the form title, that would really be an unhelpful place to hide that setting.

I think the best place would actually be in **Administer > System Settings > Outbound Email**:

- It's available even without CiviMail.
- It's clearly mail-related.
- There's a reasonable chance you're in there anyway setting up SMTP if you are the kind who wants to send hundreds of messages without using CiviMail.

**Comment by agh1** [comment]
If you do add this feature, I also think "Max emails kill switch" is a very confusing name.  If told that setting name, I'd think it was a mechanism to halt the system from sending the emails currently going out.

Again, I'm still not endorsing the idea, but if you do this, I'd propose the title be "Quick email recipient limit".  We lack a good name for the non-CiviMail "Send an email" task, but I've noticed people saying "quick email" and kind of like it.

**Comment by eileenmcnaughton** [comment]
@agh1 there are spammers and there are inadvertent spammers. We can't stop spammers but we do consider the option of simply not exposing through the UI to prevent inadvertent spammers? The issue is of course that people may not be educated enough on the implications of what they are doing (or the fact that in the US they can be fined $40,000 per email if they get it wrong)

I'm inclined to move the UI out of the existing PR & merge that & resolve the UI in a separate PR (It's currently a one-line change to add it to the UI although moving it to a new form might make it bigger).

**Comment by eileenmcnaughton** [comment]
@agh1 the discussion is on this PR but actually we are probably going to merge #10705 - which has different wording

**Comment by eileenmcnaughton** [comment]
@agh1 Just to be clear I am 100% happy with adding this setting without exposing it anywhere in the UI (other than api explorer I guess) - if that is your preference then I would argue it's easier to leave it off now & add it later than the opposite

**Comment by agh1** [comment]
@eileenmcnaughton I think that's a fabulous plan: add it as a setting without a UI.  This makes it hard for the bumbling tinkerer to change it, easy for the responsible expert to change it, and allows @omarabuhussein to release the uk.co.compucorp.spamalot extension to expose the field wherever he likes.  LExIM!

**Comment by eileenmcnaughton** [comment]
Lol

**Comment by omarabuhussein** [comment]
interesting , I had no idea that sending spam from a product will affect the product too and not just the domain (or ip) that sends these emails.

anyway , if this going to be a hidden setting , why not just allow it to be added to civicrm.settings.php file and read it from there (or use the default 50 if it is not set) instead of this https://github.com/civicrm/civicrm-core/pull/10705  ? wouldn't that be a more hidden place but still easier for experts to find and configure ?

**Comment by eileenmcnaughton** [comment]
@omarabuhussein I think #10705 now does what you are saying?

**Comment by eileenmcnaughton** [comment]
Re giving the product a bad name - obviously if we leave no fingerprint or markers on the email then we are safe from those markers being counted as spam markers - I'm not 100% sure that is the case, or if it is if it always will be

**Comment by eileenmcnaughton** [comment]
@omarabuhussein I just checked back on this & my feeling is it is good to merge except for the change @colemanw identified & requested (which is minor & relates to improving the construction of a string for translation). If you can make that change I believe this is mergeable

**Comment by eileenmcnaughton** [comment]
hang on - there are 2 prs - now I need to check which is the one to merge!

**Comment by eileenmcnaughton** [comment]
I merged the other & am closing this. The merge-ready approach turned out to have the opposite effect than intended. ie. rather than giving time to reflect it made the discussion go cold & nearly had the effect of the wrong PR being merged.

**Pull request closed without merging**